cli/src/commands/operation.rs

```diff
@@ -395,14 +460,541 @@ fn cmd_op_abandon(
+        if merged_parents_op.is_none() {
+            return Err(user_error("Cannot diff operation with no parents"));
+        }
+        from_op = merged_parents_op.unwrap();
+    }
+    let with_content_format = LogContentFormat::new(ui, command.settings())?;
+
+    let from_repo = repo_loader.load_at(&from_op)?;
+    let to_repo = repo_loader.load_at(&to_op)?;
+
+    ui.request_pager();
+    writeln!(
+        ui.stdout(),
+        "From operation {}: {}",
+        short_operation_hash(from_op.id()),
+        from_op.metadata().description,
+    )?;
+    writeln!(
+        ui.stdout(),
+        "  To operation {}: {}",
+        short_operation_hash(to_op.id()),
+        to_op.metadata().description,
+    )?;
+    writeln!(ui.stdout())?;
+
+    show_op_diff(
+        ui,
+        command,
+        &from_repo,
+        &to_repo,
+        !args.no_graph,
+        &with_content_format,
+        &args.diff_format,
+        args.patch,
+    )
+}
+
+// Merges the given `operations` into a single operation. Returns `None` if
+// there are no operations to merge.
+fn merge_operations(
+    repo_loader: &RepoLoader,
+    settings: &UserSettings,
+    mut operations: impl ExactSizeIterator<Item = OpStoreResult<Operation>>,
+) -> Result<Option<Operation>, CommandError> {
+    let num_operations = operations.len();
+    if num_operations == 0 {
+        return Ok(None);
+    }
+
+    let base_op = operations.next().transpose()?.unwrap();
+    let final_op = if num_operations > 1 {
+        let base_repo = repo_loader.load_at(&base_op)?;
+        let mut tx = base_repo.start_transaction(settings);
+        for other_op in operations {
+            let other_op = other_op?;
+            tx.merge_operation(other_op)?;
+            tx.mut_repo().rebase_descendants(settings)?;
+        }
+        let tx_description = format!("merge {} operations", num_operations);
+        let merged_repo = tx.write(tx_description).leave_unpublished();
+        merged_repo.operation().clone()
+    } else {
+        base_op
+    };
+
+    Ok(Some(final_op))
+}
+
+// Computes and shows the differences between two operations, using the given
+// `Repo`s for the operations.
+#[allow(clippy::too_many_arguments)]
+fn show_op_diff(
+    ui: &mut Ui,
+    command: &CommandHelper,
+    from_repo: &Arc<ReadonlyRepo>,
+    to_repo: &Arc<ReadonlyRepo>,
+    show_graph: bool,
+    with_content_format: &LogContentFormat,
+    diff_format_args: &DiffFormatArgs,
+    patch: bool,
+) -> Result<(), CommandError> {
+    let diff_workspace_command =
+        command.for_loaded_repo(ui, command.load_workspace()?, to_repo.clone())?;
+    let diff_renderer = diff_workspace_command.diff_renderer_for_log(diff_format_args, patch)?;
+
+    // Create a new transaction starting from `to_repo`.
+    let mut workspace_command =
+        command.for_loaded_repo(ui, command.load_workspace()?, to_repo.clone())?;
+    let mut tx = workspace_command.start_transaction();
```

Comment:
It might be better to use low-level `repo.start_transaction()`. `WorkspaceCommandHelper` assumes that the loaded repo is synchronized with the working copy, and things might get weird if we committed the transaction by mistake.

Reply:
Hmm, I'm depending on `WorkspaceCommandTransaction` for `write_commit_summary` though. Would it be possible to reuse the existing code in some way?